spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/core/cloudfoundry/GcpCloudFoundryEnvironmentPostProcessor.java

```diff
@@ -0,0 +1,153 @@
+				"subscription_name", "subscription-name")),
+		storage("google-storage",
+				ImmutableMap.of("ProjectId", "project-id",
+				"PrivateKeyData", "credentials.encodedKey",
+				"bucket_name", "bucket-name")),
+		spanner("google-spanner",
+				ImmutableMap.of("ProjectId", "project-id",
+						"PrivateKeyData", "credentials.encodedKey",
+						"instance_id", "instance-id")),
+		trace("google-stackdriver-trace",
+				ImmutableMap.of("ProjectId", "project-id",
+						"PrivateKeyData", "credentials.encodedKey"));
+
+		/**
+		 * Name of the GCP Cloud Foundry service in the VCAP_SERVICES JSON.
+		 */
+		private String cfServiceName;
+
+		/**
+		 * Direct mapping of GCP service broker field names in VCAP_SERVICES JSON to Spring Cloud
+		 * GCP property names. {@link #retrieveCfProperties(Map, GcpCfService)} uses this map to
+		 * perform the actual transformation.
+		 *
+		 * <p>For instance, "ProjectId" for the "google-storage" service will map to
+		 * "spring.cloud.gcp.storage.project-id" field.</p>
+		 */
+		private Map<String, String> cfPropNameToGcp;
+
+		GcpCfService(String cfServiceName,
+				Map<String, String> cfPropNameToGcp) {
+			this.cfServiceName = cfServiceName;
+			this.cfPropNameToGcp = cfPropNameToGcp;
+		}
+
+		public String getCfServiceName() {
+			return this.cfServiceName;
+		}
+
+		public Map<String, String> getCfPropNameToGcp() {
+			return this.cfPropNameToGcp;
+		}
+	}
+
+	@Override
+	public void postProcessEnvironment(ConfigurableEnvironment environment,
+			SpringApplication application) {
+		if (!StringUtils.isEmpty(environment.getProperty(VCAP_SERVICES_ENVVAR))) {
+			Map<String, Object> vcapMap =
+					this.parser.parseMap(environment.getProperty(VCAP_SERVICES_ENVVAR));
+
+			Properties gcpCfPubSubProps = new Properties();
+
+			Arrays.stream(GcpCfService.values()).forEach(cfService ->
+					gcpCfPubSubProps.putAll(retrieveCfProperties(vcapMap, cfService)));
+
+			environment.getPropertySources()
+					.addFirst(new PropertiesPropertySource("gcpCf", gcpCfPubSubProps));
+		}
+	}
+
+	private static Properties retrieveCfProperties(Map<String, Object> vcapMap,
+			GcpCfService service) {
+		Properties properties = new Properties();
+		List<Object> serviceBindings = (List<Object>) vcapMap.get(service.getCfServiceName());
+
+		if (serviceBindings == null) {
+			return properties;
+		}
+
+		if (serviceBindings.size() != 1) {
+			LOGGER.info("The service " + service.getCfServiceName() + " has to be bound to a "
+					+ "Cloud Foundry application once and only once.");
+			return properties;
+		}
+
+		Map<String, Object> serviceBinding = (Map<String, Object>) serviceBindings.get(0);
+		Map<String, String> credentialsMap = (Map<String, String>) serviceBinding.get("credentials");
+		String prefix = SPRING_CLOUD_GCP_PROPERTY_PREFIX + service.name() + ".";
+		service.getCfPropNameToGcp().forEach(
+				(cfPropKey, gcpPropKey) -> properties.put(
+						prefix + gcpPropKey,
+						credentialsMap.get(cfPropKey)));
```

Comment:
Is there some confirmation that `topic_name` or `topic-name` really maps to the `GcpPubSubProperties.defaultTopicName` ?

Thanks

Reply:
Good catch!